pylint custom configuration overrides plugin defaults settings.

pylint overrides default configuration set by plugins using parameters set by rc file or command line arguments:
https://github.com/PyCQA/pylint/blob/fcc01516ae176ad3fdedc4497328105f3314e376/pylint/lint.py#L1568-L1571

pylint-django extends good-names in `register()` method [1]:
```python
    name_checker = get_checker(linter, NameChecker)
    name_checker.config.good_names += ('qs', 'urlpatterns', 'register', 'app_name', 'handler500')
    # we don't care about South migrations
    linter.config.black_list += ('migrations', 'south_migrations')
```

Currently plugin is not able to avoid overriding of configuration. pylint should provide a hook for plugins which is executed *after* configuration is set. Currently there are multiple open issues in pylint-django project affected by this defect: 
* https://github.com/PyCQA/pylint-django/issues/181 
* https://github.com/PyCQA/pylint-django/issues/167

### Steps to reproduce
1. Install django and pylint-django using pip
2. Create empty project: `django-admin startproject myproject`
3. Execute pylint with pylint-django plugin and with default configuration:
```
pylint --load-plugins=pylint_django myproject/myproject/urls.py

------------------------------------
Your code has been rated at 10.00/10
```
4. Execute pylint with django-pylint and custom configuration:
```
pylint --load-plugins=pylint_django --good-names=foo,bar  myproject/myproject/urls.py

************* Module myproject.urls
myproject/myproject/urls.py:19:0: C0103: Constant name "urlpatterns" doesn't conform to UPPER_CASE naming style (invalid-name)

-------------------------------------------------------------------
Your code has been rated at 6.67/10 (previous run: 10.00/10, -3.33)
```
### Current behavior
Currently, configuration provided by pylint-django is overriden by custom configuration (see [1] pylint-django configuration)

### Expected behavior
Custom configuration should not override pylint-django customization

### pylint --version output
pylint 2.1.1
astroid 2.0.4
Python 3.6.4 (default, Jan  7 2018, 15:53:53)
[GCC 6.4.0]

[1] https://github.com/PyCQA/pylint-django/blob/c5b5bfdef66453575074b36018ee716411bcb0a4/pylint_django/plugin.py#L18-L22

--- a/CONTRIBUTORS.txt
+++ b/CONTRIBUTORS.txt
@@ -255,3 +255,5 @@ contributors:
 
 * Pablo Galindo Salgado: contributor
   Fix false positive 'Non-iterable value' with async comprehensions.
+
+* Matus Valo

--- a/ChangeLog
+++ b/ChangeLog
@@ -7,6 +7,16 @@ What's New in Pylint 2.3.0?
 
 Release date: TBA
 
+* Added ``load_configuration()`` hook for plugins
+
+  New optional hook for plugins is added: ``load_configuration()``.
+  This hook is executed after configuration is loaded to prevent
+  overwriting plugin specific configuration via user-based
+  configuration.
+
+  Close #2635
+
+
 * Add ``no-else-raise`` warning (R1720)
 
   Close #2558

--- a/doc/how_tos/plugins.rst
+++ b/doc/how_tos/plugins.rst
@@ -53,7 +53,7 @@ We can extend hello-world plugin to ignore some specific names using
     name_checker.config.good_names += ('Hello', 'World')
 
     # We ignore bin directory
-    linter.config.black_list += ('bin')
+    linter.config.black_list += ('bin',)
 
 Depending if we need a **transform plugin** or a **checker**, this might not
 be enough. For the former, this is enough to declare the module as a plugin,

--- a/doc/how_tos/plugins.rst
+++ b/doc/how_tos/plugins.rst
@@ -14,6 +14,12 @@ and tailored to a particular module, library of framework.
 In general, a plugin is a module which should have a function ``register``,
 which takes an instance of ``pylint.lint.PyLinter`` as input.
 
+A plugin can optionally define also function ``load_configuration``,
+which takes an instance of ``pylint.lint.PyLinter`` as input. This
+function is called after Pylint loads configuration from configuration
+file and command line interface. This function should load additional
+plugin specific configuration to Pylint.
+
 So a basic hello-world plugin can be implemented as:
 
 .. sourcecode:: python
@@ -22,6 +28,7 @@ So a basic hello-world plugin can be implemented as:
   def register(linter):
     print 'Hello world'
 
+
 We can run this plugin by placing this module in the PYTHONPATH and invoking
 **pylint** as:
 
@@ -29,6 +36,24 @@ We can run this plugin by placing this module in the PYTHONPATH and invoking
 
   $ pylint -E --load-plugins hello_plugin foo.py
   Hello world
+
+We can extend hello-world plugin to ignore some specific names using
+``load_configuration`` function:
+
+.. sourcecode:: python
+
+  # Inside hello_plugin.py
+  def register(linter):
+    print 'Hello world'
+
+  def load_configuration(linter):
+
+    name_checker = get_checker(linter, NameChecker)
+    # We consider as good names of variables Hello and World
+    name_checker.config.good_names += ('Hello', 'World')
+
+    # We ignore bin directory
+    linter.config.black_list += ('bin')
 
 Depending if we need a **transform plugin** or a **checker**, this might not
 be enough. For the former, this is enough to declare the module as a plugin,

--- a/pylint/lint.py
+++ b/pylint/lint.py
@@ -279,6 +279,10 @@ if multiprocessing is not None:
                 linter.load_plugin_modules(self._plugins)
 
             linter.load_configuration_from_config(self._config)
+
+            # Load plugin specific configuration
+            linter.load_plugin_configuration()
+
             linter.set_reporter(reporters.CollectingReporter())
 
             # Enable the Python 3 checker mode. This option is
@@ -649,6 +653,16 @@ class PyLinter(
             self._dynamic_plugins.add(modname)
             module = modutils.load_module_from_name(modname)
             module.register(self)
+
+    def load_plugin_configuration(self):
+        """Take a list of module names which ar pylint plugins and
+        calls configuration hook function to load their specific
+        configuration
+        """
+        for modname in self._dynamic_plugins:
+            module = modutils.load_module_from_name(modname)
+            if hasattr(module, 'load_configuration'):
+                module.load_configuration(self)
 
     def _load_reporter(self):
         name = self._reporter_name.lower()
@@ -1569,6 +1583,7 @@ group are mutually exclusive.",
         # now we can load file config and command line, plugins (which can
         # provide options) have been registered
         linter.load_config_file()
+
         if reporter:
             # if a custom reporter is provided as argument, it may be overridden
             # by file parameters, so re-set it here, but before command line
@@ -1601,6 +1616,10 @@ group are mutually exclusive.",
             else:
                 if linter.config.jobs == 0:
                     linter.config.jobs = _cpu_count()
+
+        # We have loaded configuration from config file and command line. Now, we can
+        # load plugin specific configuration.
+        linter.load_plugin_configuration()
 
         # insert current working directory to the python path to have a correct
         # behaviour

--- a/pylint/lint.py
+++ b/pylint/lint.py
@@ -661,7 +661,7 @@ class PyLinter(
         """
         for modname in self._dynamic_plugins:
             module = modutils.load_module_from_name(modname)
-            if hasattr(module, 'load_configuration'):
+            if hasattr(module, "load_configuration"):
                 module.load_configuration(self)
 
     def _load_reporter(self):

--- a/pylint/test/regrtest_data/dummy_plugin/dummy_conf_plugin.py
+++ b/pylint/test/regrtest_data/dummy_plugin/dummy_conf_plugin.py
@@ -1,0 +1,6 @@
+def register(linter):
+    pass
+
+
+def load_configuration(linter):
+    linter.config.black_list += ('bin',)

--- a/pylint/test/unittest_lint.py
+++ b/pylint/test/unittest_lint.py
@@ -511,6 +511,56 @@ def test_addmessage_invalid(linter):
     assert str(cm.value) == "Message C0321 must provide Node, got None"
 
 
+def test_load_plugin_command_line():
+    dummy_plugin_path = join(HERE, "regrtest_data", "dummy_plugin")
+    sys.path.append(dummy_plugin_path)
+
+    run = Run(
+        ["--load-plugins", "dummy_plugin", join(HERE, "regrtest_data", "empty.py")],
+        do_exit=False,
+    )
+    assert (
+        len([ch.name for ch in run.linter.get_checkers() if ch.name == "dummy_plugin"])
+        == 2
+    )
+
+    sys.path.remove(dummy_plugin_path)
+
+
+def test_load_plugin_config_file():
+    dummy_plugin_path = join(HERE, "regrtest_data", "dummy_plugin")
+    sys.path.append(dummy_plugin_path)
+    config_path = join(HERE, "regrtest_data", "dummy_plugin.rc")
+
+    run = Run(
+        ["--rcfile", config_path, join(HERE, "regrtest_data", "empty.py")],
+        do_exit=False,
+    )
+    assert (
+        len([ch.name for ch in run.linter.get_checkers() if ch.name == "dummy_plugin"])
+        == 2
+    )
+
+    sys.path.remove(dummy_plugin_path)
+
+
+def test_load_plugin_configuration():
+    dummy_plugin_path = join(HERE, "regrtest_data", "dummy_plugin")
+    sys.path.append(dummy_plugin_path)
+
+    run = Run(
+        [
+            "--load-plugins",
+            "dummy_conf_plugin",
+            "--ignore",
+            "foo,bar",
+            join(HERE, "regrtest_data", "empty.py"),
+        ],
+        do_exit=False,
+    )
+    assert run.linter.config.black_list == ["foo", "bar", "bin"]
+
+
 def test_init_hooks_called_before_load_plugins():
     with pytest.raises(RuntimeError):
         Run(["--load-plugins", "unexistant", "--init-hook", "raise RuntimeError"])

